Fix experiment id resolution in projects

drewmcdonald opened the pull request:
## What changes are proposed in this pull request?
 
Addresses #1353. 

- Fixes bug where we try to access `experiment.experiment_id` before knowing `experiment` exists
- We now create an experiment if the given experiment name is not found, which is the behavior in other parts of MLflow (see tracking.fluent.set_experiment)
 
## How is this patch tested?
 
Simplified and extended unit tests for `_resolve_experiment_id`.
 
## Release Notes
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
Passing `--experiment-name` to a project run will create the experiment if it doesn't exist, matching the behavior in the fluent api.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [x] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [x] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

Commit:
clean up experiment id resolution in projects usage:
- don't access experiment.experiment_id before knowing experiment exists
- create new experiment if experiment_name doesn't exist

sueann commented on mlflow/projects/__init__.py:
```diff
@@ -67,12 +70,19 @@ def _resolve_experiment_id(experiment_name=None, experiment_id=None):
+            return exp.experiment_id
+        else:
+            print("INFO: '{}' does not exist. Creating a new experiment".format(experiment_name))
+            return client.create_experiment(experiment_name)
```
you want to return the experiment ID right?

sueann replied:
It'd be worth adding some simple unit tests for this function so we can catch things like this. Maybe in https://github.com/mlflow/mlflow/blob/master/tests/projects/test_projects_cli.py. `client.create_experiment` can be mocked.

drewmcdonald replied:
Will do! I think `client.create_experiment` returns the created ID, but we'll find out for sure when I do the tests :)

sueann replied:
Ah my bad - you're right it returns the ID. Probably still a good idea to add tests to ensure the different cases work though. Thank you!

sueann commented on mlflow/R/mlflow/R/install.R:
```diff
@@ -10,11 +10,14 @@ mlflow_conda_env_name <- function() {
 # Create conda env used by MLflow if it doesn't already exist
 #' @importFrom reticulate conda_install conda_create conda_list
-mlflow_maybe_create_conda_env <- function() {
+#' @param python_version Python version to use within conda environment created for
```
Is this python_version change related to the experiment_id fix? Could you explain these changes in the PR description or move to a different PR? Thanks!

drewmcdonald replied:
sorry, those commits were pulled from the patch to get builds working again. removed since they're not related to this change

sueann reviewed (commented):
Thanks for the fix! A few comments for you.

Commit:
make resolution pattern more explicit

Commit:
docstring backticks

Commit:
cast final experiment_id to str:
- matches client.get_experiment_by_name
- matches client.create_experiment

Commit:
experiment id resolution tests:
- move tests to test_projects_cli.py
- simplify tests for successful experiment id resolution
- add test cases

Commit:
lint

drewmcdonald commented:
@sueann I've simplified and extended the tests for `_resolve_experiment_id` and added backticks in the docstring.

One catch from writing the tests is that sources of `experiment_id` return strings except (potentially) the directly passed `experiment_id`. I'm coercing that to a string for now -- if at some point one of the other resolution methods switches to an int the tests should fail and spur someone to troubleshoot further.

sueann commented on tests/projects/test_projects_cli.py:
```diff
@@ -103,3 +105,24 @@ def test_run_databricks_cluster_spec(tmpdir):
         assert res.exit_code != 0
+
+
+@pytest.mark.parametrize('experiment_name,experiment_id,expected', [
```
er, i didn't realize these lived in test_projects.py! sorry do you mind moving them back (with your added test cases of course). please feel free to call me out if i don't make full sense :-) thanks!

drewmcdonald replied:
Cool should be good to go!

sueann commented on mlflow/projects/__init__.py:
```diff
@@ -64,14 +64,14 @@ def _resolve_experiment_id(experiment_name=None, experiment_id=None):
     :param experiment_name: Name of experiment under which to launch the run.
     :param experiment_id: ID of experiment under which to launch the run.
-    :return: int
+    :return: str
```
thanks! yeah experiment_id is meant to be string now. (this particular description: https://github.com/mlflow/mlflow/blob/80099417284f9c7377eb8af58b2845291abdea62/mlflow/tracking/fluent.py#L287 is wrong so it is confusing)

Commit:
move `_resolve_experiment` tests back to test_projects.py

Commit:
lint

sueann commented on tests/projects/test_projects_cli.py:
```diff
@@ -13,7 +13,6 @@
 from tests.integration.utils import invoke_cli_runner
 from tests.projects.utils import TEST_PROJECT_DIR, GIT_PROJECT_URI, SSH_PROJECT_URI, \
     TEST_NO_SPEC_PROJECT_DIR
-from tests.projects.utils import tracking_uri_mock  # pylint: disable=unused-import
```
i think this is still needed

drewmcdonald replied:
trusted my ide too much, sigh.

sueann commented on mlflow/projects/__init__.py:
```diff
@@ -59,20 +59,30 @@ def _resolve_experiment_id(experiment_name=None, experiment_id=None):
 
     Verifies either one or other is specified - cannot be both selected.
 
+    If `experiment_name` is provided and does not exist, an experiment
```
double-backticks please (it's hard to convey this in github comments because it tries to format them... aha,
```
``experiment_name``
```
is what we need

sueann reviewed (commented):
looks like an import error in test failure (+ a nit) - otherwise lgtm. thanks!

Commit:
double backticks

Commit:
restore dropped unused-but-not-really import

drewmcdonald commented:
Hmmm build is failing but looks like something unrelated to this PR

sueann commented:
Yeah - someone is looking into fixing the tests. I'll hold off on merging this until that is done just in case. Thanks!
